pkg/yurthub/kubernetes/meta/restmapper.go

```diff
@@ -0,0 +1,259 @@
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/klog"
+)
+
+const (
+	CacheDynamicRESTMapperKey = "_internal/restmapper/cache-crd-restmapper.conf"
+	SepForGVR                 = "/"
+)
+
+var (
+	// unsafeSchemeRESTMapper is used to store the mapping relationship between GVK and GVR in scheme
+	// It is not updatable and is only used for the judgment of scheme resources
+	unsafeSchemeRESTMapper = NewDefaultRESTMapperFromScheme()
+	ErrGVRNotRecognized    = errors.New("GroupVersionResource is not recognized")
+)
+
+// RESTMapperManager is responsible for managing different kind of RESTMapper
+type RESTMapperManager struct {
+	sync.RWMutex
+	storage storage.Store
+	// UnsafeDefaultRESTMapper is used to save the GVK and GVR mapping relationships of built-in resources
+	unsafeDefaultRESTMapper *meta.DefaultRESTMapper
+	// dynamicRESTMapper is used to save the GVK and GVR mapping relationships of Custom Resources
+	dynamicRESTMapper map[schema.GroupVersionResource]schema.GroupVersionKind
+}
+
+func NewDefaultRESTMapperFromScheme() *meta.DefaultRESTMapper {
+	s := scheme.Scheme
+	defaultGroupVersions := s.PrioritizedVersionsAllGroups()
+	mapper := meta.NewDefaultRESTMapper(defaultGroupVersions)
+	// enumerate all supported versions, get the kinds, and register with the mapper how to address
+	// our resources.
+	for _, gv := range defaultGroupVersions {
+		for kind := range s.KnownTypes(gv) {
+			// Only need to process non-list resources
+			if !strings.HasSuffix(kind, "List") {
+				// Since RESTMapper is only used for mapping GVR to GVK information,
+				// the scope field is not involved in actual use,
+				// so all scope are currently set to meta.RESTScopeNamespace
+				scope := meta.RESTScopeNamespace
+				mapper.Add(gv.WithKind(kind), scope)
+			}
+		}
+	}
+	return mapper
+}
+
+func NewRESTMapperManager(rootDir string) *RESTMapperManager {
+	var dm map[schema.GroupVersionResource]schema.GroupVersionKind
+	dStorage, err := disk.NewDiskStorage(rootDir)
+	if err == nil {
+		// Recover the mapping relationship between GVR and GVK from the hard disk
+		b, err := dStorage.Get(CacheDynamicRESTMapperKey)
+		if err == nil && len(b) != 0 {
+			dm = unmarshalDynamicRESTMapper(b)
+			klog.Infof("reset DynamicRESTMapper to %v", dm)
+		} else {
+			dm = make(map[schema.GroupVersionResource]schema.GroupVersionKind)
+			klog.Infof("initialize an empty DynamicRESTMapper")
+		}
+	} else {
+		klog.Errorf("failed to create disk storage, %v, only initialize an empty DynamicRESTMapper in memory ", err)
+	}
+
+	return &RESTMapperManager{
+		unsafeDefaultRESTMapper: unsafeSchemeRESTMapper,
+		dynamicRESTMapper:       dm,
+		storage:                 dStorage,
+	}
+}
+
+// Obtain gvk according to gvr in dynamicRESTMapper
+func (rm *RESTMapperManager) dynamicKindFor(gvr schema.GroupVersionResource) (schema.GroupVersionKind, error) {
+	rm.RLock()
+	defer rm.RUnlock()
```

Comment:
```
	rm.RLock()
	defer rm.RUnlock()
```

code block can be moved to before `if hasVersion` condition.

Reply:
The code is updated, and code block is removed.